File: frontend/src/types/investigation.ts
export type Verdict = 'APPROVED' | 'DECLINED' | 'NEEDS_REVIEW' | null;

export type CaseStatus =
  | 'PENDING'
  | 'INVESTIGATING'
  | 'COMPLETED'
  | 'CLOSED'
  | 'UNDER_INVESTIGATION'
  | 'VERIFICATION_PENDING'
  | 'UNRESOLVED'
  | 'CONFIRMED_FRAUD'
  | 'CLEARED'
  | 'NEEDS_REVIEW';

export interface Case {
  case_id: string;
  status: CaseStatus;
  verdict: Verdict;
  fraud_probability: number | null;
  pattern: string | null;
  exposure: number | null;
  created_at: string;
  updated_at: string;
  notes?: string | null;
  customer_id?: string | null;
}

export interface EvidenceItem {
  id?: string;
  evidence_id?: string;
  type?: string;
  evidence_type?: string;
  source?: string;
  description?: string;
  related_entity?: string | null;
  transaction_id?: string | null;
  card_id?: string | null;
  strength?: string | null;
  raw_data?: Record<string, any>;
  details?: Record<string, any>;
  risk_signal?: string | number | null;
  timestamp?: string | null;
}

export interface EvidenceRequest {
  request_id: string;
  case_id?: string;
  transaction_id?: string | null;
  request_type?: string;
  request_text?: string | null;
  status: string; // PENDING | RESPONDED | CANCELLED
  response?: string | null;
  response_source?: string | null;
  response_assumptions?: string | null;
  responded_at?: string | null;
  cancelled_reason?: string | null;
  cancelled_at?: string | null;
  triggered_investigation_id?: string | null;
  actor?: string | null;
  created_at?: string;
  updated_at?: string;
  notes?: string | null;
  // Backward-compat aliases from TigerGraph EvidenceRequestResult
  details?: Record<string, any>;
  requested_at?: string | null;
}

// ---- Payloads for API calls ----

export interface EvidenceRequestCreatePayload {
  transaction_id?: string | null;
  request_type?: string;
  request_text: string;
  notes?: string | null;
  actor?: string;
}

export interface EvidenceRequestRespondPayload {
  response: string;
  response_source?: string;
  response_assumptions?: string | null;
  actor?: string;
}

export interface EvidenceRequestCancelPayload {
  cancelled_reason?: string | null;
  actor?: string;
}

export interface EvidenceRequestRespondResult {
  evidence_request: EvidenceRequest;
  new_investigation_id: string | null;
  investigation_triggered: boolean;
  investigation_error: string | null;
}

export interface KeyEvidenceFinding {
  evidence_id: string;
  finding: string;
  significance: 'LOW' | 'MEDIUM' | 'HIGH' | 'NEUTRAL' | string;
}

export interface ReasoningResult {
  summary?: string;
  key_evidence?: KeyEvidenceFinding[];
  observed_patterns?: string[];
  conflicting_evidence?: string[];
  missing_evidence?: string[];
  uncertainties?: string[];
  relevant_rules?: string[];
  reasoning?: string;
}

export interface InvestigationResult {
  case_id: string;
  customer_id?: string | null;
  case_status: CaseStatus;
  status: CaseStatus;
  verdict: Verdict;
  fraud_probability: number | null;
  pattern: string | null;
  evidence: EvidenceItem[];
  affected_transaction_ids: string[];
  connected_card_ids: string[];
  connected_device_ids: string[];
  exposure: number;
  similar_prior_cases: string[];
  written_to_graph: boolean;
  evidence_requests: EvidenceRequest[];
  next_best_actions_initial: string[];
  next_best_actions_final: string[];
  rules_evaluated: RuleEvaluationItem[];
  SAR?: {
    status?: string;
    reason?: string;
    [key: string]: any;
  } | null;
  stop_reason?: string | null;
  tool_calls?: Record<string, any>[];
  tokens?: { prompt: number; completion: number; total: number };
  latency?: number;
  evidence_count?: number;
  reasoning_summary?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface HealthCheckResponse {
  status: string;
}

export interface InvestigationHistoryItem {
  investigation_id: string;
  case_id: string;
  customer_id?: string | null;
  case_status?: string | null;
  status?: string | null;
  verdict?: Verdict;
  created_at: string;
  completed_at?: string | null;
  reasoning_summary?: string | null;
  stop_reason?: string | null;
}

export interface InvestigationHistoryResponse {
  case_id: string;
  investigations: InvestigationHistoryItem[];
}

export interface AuditEventItem {
  id: number;
  case_id: string;
  investigation_id?: string | null;
  event_type: string;
  description?: string | null;
  actor: 'SYSTEM' | 'AGENT' | 'ANALYST' | string;
  metadata_json?: string | null;
  created_at: string;
}

export interface RuleEvaluationItem {
  rule_id: string;
  rule_name?: string;
  triggered: boolean;
  reason?: string;
  description?: string;
  severity?: string;
  created_at?: string;
}

export interface InvestigationDetailResponse {
  investigation_id: string;
  case_id: string;
  customer_id?: string | null;
  case_status?: CaseStatus;
  status?: CaseStatus;
  verdict?: Verdict;
  fraud_probability?: number | null;
  pattern?: string | null;
  reasoning_summary?: string | null;
  stop_reason?: string | null;
  exposure?: number | null;
  sar_status?: string | null;
  sar_reason?: string | null;
  llm_provider?: string | null;
  llm_model?: string | null;
  llm_latency?: number | null;
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  total_tokens?: number | null;
  created_at: string;
  completed_at?: string | null;
  evidence: EvidenceItem[];
  actions_initial: string[];
  actions_final: string[];
  rules: RuleEvaluationItem[];
  audit_events: AuditEventItem[];
}

export interface GraphNode {
  id: string;
  type: 'Customer' | 'Card' | 'Transaction' | 'DeviceProfile' | 'EmailDomain' | 'BillingRegion' | 'ClosedCase' | 'EvidenceRequest' | string;
  label: string;
  properties: Record<string, any>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  type: 'OWNS' | 'MADE' | 'FROM_DEVICE' | 'PURCHASER_EMAIL' | 'BILLED_IN' | 'INVOLVES' | 'CONNECTED_TO' | 'FOR_CASE' | 'FOR_TRANSACTION' | string;
  label: string;
}

export interface CaseGraphResponse {
  case_id: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
}

File: backend/app/schemas/investigation.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from app.agent.schemas import (
    EvidenceItem,
    InvestigationRequest,
    InvestigationResult
)

# ---------------------------------------------------------------------------
# Evidence Request lifecycle schemas
# ---------------------------------------------------------------------------

class EvidenceRequestCreate(BaseModel):
    """Payload for creating a new evidence request."""
    transaction_id: Optional[str] = None
    request_type: str = "customer_verification"
    request_text: str = Field(..., min_length=5, description="The question / verification text sent to the customer.")
    notes: Optional[str] = None
    actor: Optional[str] = "ANALYST"


class EvidenceRequestRespond(BaseModel):
    """Payload for recording an actual customer / analyst response to a PENDING request."""
    response: str = Field(..., min_length=1, description="The actual response text received from the customer or analyst.")
    response_source: str = Field(default="CUSTOMER", description="Who provided the response: CUSTOMER, ANALYST, or SYSTEM.")
    response_assumptions: Optional[str] = None
    actor: Optional[str] = "ANALYST"


class EvidenceRequestCancel(BaseModel):
    """Payload for cancelling a PENDING evidence request."""
    cancelled_reason: Optional[str] = None
    actor: Optional[str] = "ANALYST"


class EvidenceRequestResponse(BaseModel):
    """API response schema for a persisted evidence request."""
    request_id: str
    case_id: str
    transaction_id: Optional[str] = None
    request_type: str
    request_text: str
    status: str  # PENDING, RESPONDED, CANCELLED
    response: Optional[str] = None
    response_source: Optional[str] = None
    response_assumptions: Optional[str] = None
    responded_at: Optional[datetime] = None
    cancelled_reason: Optional[str] = None
    cancelled_at: Optional[datetime] = None
    triggered_investigation_id: Optional[str] = None
    actor: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    notes: Optional[str] = None

    class Config:
        from_attributes = True


class EvidenceRequestRespondResult(BaseModel):
    """Extended response returned when a respond action also triggers a new investigation."""
    evidence_request: EvidenceRequestResponse
    new_investigation_id: Optional[str] = None
    investigation_triggered: bool = False
    investigation_error: Optional[str] = None

class HealthCheck(BaseModel):
    status: str = "ok"

class CaseBase(BaseModel):
    case_id: str
    customer_id: Optional[str] = None
    status: str
    verdict: Optional[str] = None
    fraud_probability: Optional[float] = None
    pattern: Optional[str] = None
    exposure: Optional[float] = None

class CaseCreate(CaseBase):
    notes: Optional[str] = None

class CaseResponse(CaseBase):
    created_at: datetime
    updated_at: datetime
    notes: Optional[str] = None

    class Config:
        from_attributes = True

class InvestigationHistoryItem(BaseModel):
    investigation_id: str
    case_id: str
    customer_id: Optional[str] = None
    case_status: Optional[str] = None
    status: Optional[str] = None
    verdict: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    reasoning_summary: Optional[str] = None
    stop_reason: Optional[str] = None

    class Config:
        from_attributes = True

class InvestigationHistoryResponse(BaseModel):
    case_id: str
    investigations: List[InvestigationHistoryItem]

class AuditEventItem(BaseModel):
    id: int
    case_id: str
    investigation_id: Optional[str] = None
    event_type: str
    description: Optional[str] = None
    actor: str = "AGENT"
    metadata_json: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True

class InvestigationDetailResponse(BaseModel):
    investigation_id: str
    case_id: str
    customer_id: Optional[str] = None
    case_status: Optional[str] = None
    status: Optional[str] = None
    verdict: Optional[str] = None
    fraud_probability: Optional[float] = None
    pattern: Optional[str] = None
    reasoning_summary: Optional[str] = None
    stop_reason: Optional[str] = None
    exposure: Optional[float] = None
    sar_status: Optional[str] = None
    sar_reason: Optional[str] = None
    llm_provider: Optional[str] = None
    llm_model: Optional[str] = None
    llm_latency: Optional[float] = None
    prompt_tokens: Optional[int] = None
    completion_tokens: Optional[int] = None
    total_tokens: Optional[int] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    evidence: List[Dict[str, Any]] = []
    actions_initial: List[str] = []
    actions_final: List[str] = []
    rules: List[Dict[str, Any]] = []
    audit_events: List[AuditEventItem] = []

    class Config:
        from_attributes = True

